src/framework/execution/__tests__/dependencyGraph.test.ts

```diff
@@ -0,0 +1,571 @@
+          resolveSpyA = resolve;
+        }),
+    );
+    const spyB = jest.fn();
+
+    /**
+     * Graph:
+     * a - b
+     *
+     * In this situation, 'a' is the leaf node
+     * 'b' depends on 'a',
+     */
+    const steps: IntegrationStep[] = [
+      {
+        id: 'a',
+        name: 'a',
+        types: [],
+        executionHandler: spyA,
+      },
+      {
+        id: 'b',
+        name: 'b',
+        types: [],
+        dependsOn: ['a'],
+        executionHandler: spyB,
+      },
+    ];
+
+    const graph = buildStepDependencyGraph(steps);
+
+    const executionPromise = executeStepDependencyGraph(graph);
+
+    // wait for spyA to have been called
+    await waitForExpect(() => {
+      expect(spyA).toHaveBeenCalledTimes(1);
+    });
+
+    // ensure spyB has not been called yet
+    expect(spyB).toHaveBeenCalledTimes(0);
+
+    resolveSpyA();
+
+    // expect spyB to eventually be called
+    await waitForExpect(() => {
+      expect(spyB).toHaveBeenCalledTimes(1);
+    });
+
+    await executionPromise;
+  });
+
+  test('should mark steps failed executionHandlers with status FAILURE a dependent steps with status PARTIAL_SUCCESS_DUE_TO_DEPENDENCY_FAILURE', async () => {
+    const spyA = jest.fn();
+    const spyB = jest.fn().mockRejectedValue(new Error('oops'));
+    const spyC = jest.fn();
+
+    /**
+     * Graph:
+     * a - b - c
+     *
+     * In this situation, 'a' is the leaf node
+     * 'b' depends on 'a',
+     * 'c' depends on 'b'
+     */
+    const steps: IntegrationStep[] = [
+      {
+        id: 'a',
+        name: 'a',
+        types: [],
+        executionHandler: spyA,
+      },
+      {
+        id: 'b',
+        name: 'b',
+        types: [],
+        dependsOn: ['a'],
+        executionHandler: spyB,
+      },
+      {
+        id: 'c',
+        name: 'c',
+        types: [],
+        dependsOn: ['b'],
+        executionHandler: spyC,
+      },
+    ];
+
+    const graph = buildStepDependencyGraph(steps);
+
+    const result = await executeStepDependencyGraph(graph);
+    expect(result).toEqual([
+      {
+        id: 'a',
+        name: 'a',
+        types: [],
+        status: IntegrationStepResultStatus.SUCCESS,
+      },
+      {
+        id: 'b',
+        name: 'b',
+        types: [],
+        dependsOn: ['a'],
+        status: IntegrationStepResultStatus.FAILURE,
+      },
+      {
+        id: 'c',
+        name: 'c',
+        types: [],
+        dependsOn: ['b'],
+        status:
+          IntegrationStepResultStatus.PARTIAL_SUCCESS_DUE_TO_DEPENDENCY_FAILURE,
+      },
+    ]);
+
+    expect(spyA).toHaveBeenCalledTimes(1);
+    expect(spyB).toHaveBeenCalledTimes(1);
+    expect(spyC).toHaveBeenCalledTimes(1);
+
+    expect(spyA).toHaveBeenCalledBefore(spyB);
+    expect(spyB).toHaveBeenCalledBefore(spyC);
+  });
+
+  test('should steps with dependencies resulting in a status PARTIAL_SUCCESS_DUE_TO_DEPENDENCY_FAILURE should also assume the same status', async () => {
```

Comment:
This is a bit hard to parse. How do you feel about this?
```suggestion
  test('steps with dependencies resulting in PARTIAL_SUCCESS_DUE_TO_DEPENDENCY_FAILURE assume the same status', async () => {
```
I generally feel the `"should "` prefix isn't very helpful since every test is declaring what should be true about the subject under test.

Reply:
engrish is hard. I like your suggestion.